### PR TITLE
Feature/topmodel

### PR DESCRIPTION
**Pull request by raztexel**
Changes for TopModel

### CODE REVIEW

File: live2live-simul.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
+import os
```
Comment (matanmarciano) [inline live2live-simul.py]: Each class needs to be in separated file. Split please...

File: live2live-simul.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    def append_filter(self, new_filter):
+        self.filter_commands.append(new_filter)
+
+
+class TargetDependentFlagsHw(TargetDependentFlags):
+    def __init__(self, fps, width, height,low_res_width, low_res_height,gop):
+        TargetDependentFlags.__init__(self, fps,gop)
+        self.decode_flags += ['-hwaccel cuvid', '-c:v h264_cuvid','-resize '+str(width)+'x'+str(height)]
+        self.filter_commands = ['fps=fps=50']
+        #self.filter_commands = self.filter_commands + ['scale_npp=in_w:in_h']
```
Comment (matanmarciano) [inline live2live-simul.py]: Remove commented code

File: live2live-simul.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    def append_filter(self, new_filter):
+        self.filter_commands.append(new_filter)
+
+
+class TargetDependentFlagsHw(TargetDependentFlags):
+    def __init__(self, fps, width, height,low_res_width, low_res_height,gop):
+        TargetDependentFlags.__init__(self, fps,gop)
+        self.decode_flags += ['-hwaccel cuvid', '-c:v h264_cuvid','-resize '+str(width)+'x'+str(height)]
+        self.filter_commands = ['fps=fps=50']
```
Comment (matanmarciano) [inline live2live-simul.py]: why 50 hard coded? provide it as parameter or minimum as constant

File: live2live-simul.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        #self.filter_commands = self.filter_commands + ['scale_npp=in_w:in_h']
+        #self.scale_command2 = 'scale_npp=' + str(low_res_width) + ':' + str(low_res_height)
+        self.video_encode_flags += ['-c:v h264_nvenc', '-preset llhq', '-b:v 15m -minrate 15m -maxrate 15m']
```
Comment (matanmarciano) [inline live2live-simul.py]: why numbers hard coded? provide it as parameter or minimum as constant

File: live2live-simul.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+class TargetDependentFlagsSw(TargetDependentFlags):
+    def __init__(self, fps,width, height,low_res_width, low_res_height,gop):
+        TargetDependentFlags.__init__(self, fps,gop)
+        self.filter_commands = self.filter_commands + ['fps=fps=' + str(fps), 'scale='+str(width)+':'+str(height)]
+        self.scale_command2 = 'scale=' + str(low_res_width) + ':' + str(low_res_height)
+        self.video_encode_flags += ['-c:v libx264']
+
+
+class File2LiveConfig:
```
Comment (matanmarciano) [inline live2live-simul.py]: I think the configuration needs to be in dedicated file and not here. Think about it.

File: live2live-simul.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    low_res_height = 100
+    loglevel = 'info'
+    jpg_encode_flags = '-vframes 1'
+    jpg_filter = 'smartblur'
+    udp_in_flag = '?pkt_size=1316?overrun_nonfatal=1&fifo_size=' + str(udp_input_fifo_size) + '"'
+    audio_encode_flags = ['-c:a aac', '-f mpegts']
+    udp_out_flag = '?pkt_size=1316?overrun_nonfatal=1&fifo_size=' + str(udp_output_fifo_size) + '"'
+
+
+    video_streams = [
```
Comment (matanmarciano) [inline live2live-simul.py]: remove commented code

File: live2live-simul.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            cmd.append(f)
+        cmd.append('-i ' + '"' + video_parameters[VideoInField.SOURCE] + configuration.udp_in_flag)
+    cmd.append(filter_complex_flags)
+    for i, video_parameters in enumerate(configuration.video_streams):
+        cmd.append('-map "[' + str(i) + '0]"')
+        for f in video_parameters[VideoInField.ENCODE_FLAG]:
+            cmd.append(f)
+        cmd.append('-map ' + str(i) + ':a -acodec copy')
+        cmd.append('"' + video_parameters[VideoInField.DESTINATION] + configuration.udp_out_flag)
+        #for f in video_parameters[VideoInField.ENCODE_FLAG]:
```
Comment (matanmarciano) [inline live2live-simul.py]: Remove commented code

File: live2live.py [hunk 8]
```diff
@@ -31,8 +31,8 @@ class ProcessField(IntEnum):
 
 
 class TargetDependentFlags:
```
Comment (matanmarciano) [inline live2live.py]: Revert changes in this file if is unnecessary

File: tTaylor.py [hunk 9]
```diff
@@ -131,13 +160,16 @@ def get_arguments(default_configuration):
...
+                        help='default is rakutan')
 
     return parser.parse_args()
 
 
 def save_cmd_to_file(t_command):
     taylor_file = open('taylor_cmd.txt', 'w')
-    taylor_file.write(t_command.replace('-map', '\n-map').replace('-pro', '\n-pro').replace('-f', '\n-f'))
+    taylor_file.write(t_command.replace('-map', '-map').replace('-pro', '-pro').replace('-f', '-f'))
     # taylor_file.write(t_command)
```
Comment (matanmarciano) [inline tTaylor.py]: remove commented code

File: tTaylor.py [hunk 10]
```diff
@@ -155,31 +187,30 @@ def save_cmd_to_file(t_command):
...
     host_destination = args.destination_ip
-    prepare_filter(args.frame_rate, args.sw_filter)
+    prepare_filter(args.frame_rate, args.sw_filter, args.mode)
     cmd = list()
     cmd.append('ffmpeg  -loglevel info')
-    for i in range(in_port_base, in_port_base + args.number_of_videos):
+    for i in reversed(range(in_port_base, in_port_base + args.number_of_videos)):
         cmd.append('-probesize '+str(args.probe_size))
-        cmd.append('-vsync 1')
+        cmd.append(' -max_delay 20000 -fflags nobuffer -analyzeduration 0 -vsync 2')
```
Comment (matanmarciano) [inline tTaylor.py]: export hard coded numbers to constants

File: tTaylor.py [hunk 11]
```diff
@@ -155,31 +187,30 @@ def save_cmd_to_file(t_command):
...
-    # cmd.append('-err_detect ignore_err')
     cmd.append('-i "logos_banner_2560x360.png"')
 
-    for i in range(audio_input_base, audio_input_base + args.number_of_audios):
-        cmd.append('-probesize '+str(args.probe_size))
-        cmd.append('-itsoffset '+args.av_offset)
-        cmd.append('-err_detect ignore_err')
-        cmd.append('-i "udp://0.0.0.0:' + str(i)+'?pkt_size=1316?overrun_nonfatal=1&fifo_size='
-                   + str(args.udp_input_fifo_size)+'"')
+    #if args.mode == 'rakutan':
```
Comment (matanmarciano) [inline tTaylor.py]: Remove commented code

File: transcodeForPet-ts.sh [hunk 12]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
```
Comment (matanmarciano) [inline transcodeForPet-ts.sh]: Think if we really need this file. there was a big risk here if the encoding command is not the same as taylor..

File: live2live-simul.py (hunk 3 above)
Comment (matanmarciano) [reply live2live-simul.py]: Still not resolved

File: live2live-simul.py (hunk 4 above)
Comment (matanmarciano) [reply live2live-simul.py]: Still not resolved

File: live2live.py (hunk 8 above)
Comment (matanmarciano) [reply live2live.py]: Still not resolved. or you past here simul content and not removed simul file?

File: live2live.py [hunk 13]
```diff
@@ -23,40 +17,29 @@ class VideoInField(IntEnum):
...
-                                    '-bf 0',
-                                    '-b:v ' + str(rate) + ' -minrate  ' + str(rate) + '  -maxrate ' + str(rate)]
+    def __init__(self, fps, width, height,low_res_width, low_res_height,gop):
+        TargetDependentFlags.__init__(self, fps,gop)
+        self.decode_flags += ['-hwaccel cuvid', '-c:v h264_cuvid','-resize '+str(width)+'x'+str(height)]
+        self.filter_commands = ['fps=fps=50']
+        self.video_encode_flags += ['-c:v h264_nvenc', '-preset llhq', '-b:v 15m -minrate 15m -maxrate 15m']
 
 
 class TargetDependentFlagsSw(TargetDependentFlags):
```
Comment (matanmarciano) [inline live2live.py]: each class need to be in separated file Raz as I commented in prevues CR

File: live2live-simul.py (hunk 4 above)
Comment (raztexel) [reply live2live-simul.py]: will be resolved by lir

File: live2live-simul.py (hunk 6 above)
Comment (raztexel) [reply live2live-simul.py]: done

File: tTaylor.py (hunk 9 above)
Comment (raztexel) [reply tTaylor.py]: done

File: transcodeForPet-ts.sh (hunk 12 above)
Comment (raztexel) [reply transcodeForPet-ts.sh]: need to be converted to python

**Comment by raztexel** [comment]
reviewed by Matan

**Pull request merged**